cmd/checkheader/main.go

```diff
@@ -0,0 +1,267 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package main
+
+import (
+	"bufio"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"os/exec"
+	"strings"
+)
+
+// The checking program implements all the checks at compile time, so there is no need to run it.
+// The generation of the checking program follows these rules:
+// - Lines are added in order of appearance.
+// - Blank lines are discarded.
+// - Comments are discarded unless they start with " #include", " #if" or " #endif".
+// - Typedefs following this pattern "typedef void* GO_%name%_PTR" are translated into "#define %name% GO_%name%_PTR".
+// - Function macros are validated against their definition in the OpenSSL headers. Example:
+//   "DEFINEFUNC(int, RAND_bytes, (uint8_t *a0, size_t a1), (a0, a1))" => "int(*__check_0)(uint8_t *, size_t) = RAND_bytes;"
+// - Function macros can be excluded when checking old OpenSSL versions by prepending '/*check:from=%version%*/', %version% being a version string such as '1.1.1' or '3.0.0'.
+
+const description = `
+Example: A check operation:
+  go run ./cmd/checkheader --ossl-include /usr/local/src/openssl-1.1.1/include ./openssl/openssl_funcs.h 
+Checkheader generates a C program, the compilation of which verifies types and functions defined in the target
+header file match the definitions in --ossl-include.
+`
+
+var osslInclude = flag.String("ossl-include", "", "OpenSSL include directory. Required.")
+var work = flag.Bool("work", false, "print the name of the temporary C program file and do not delete it when exiting.")
+
+func main() {
+	flag.Usage = func() {
+		fmt.Fprintf(flag.CommandLine.Output(), "\nUsage:\n")
+		flag.PrintDefaults()
+		fmt.Fprintf(flag.CommandLine.Output(), "%s\n\n", description)
+	}
+	flag.Parse()
+	if flag.NArg() != 1 {
+		flag.Usage()
+		os.Exit(1)
+	}
+	if *osslInclude == "" {
+		fmt.Fprintln(flag.CommandLine.Output(), "required flag not provided: --ossl-include")
+		flag.Usage()
+		os.Exit(1)
+	}
+	if _, err := os.Stat(*osslInclude); err != nil {
+		log.Fatalf("OpenSSL include directory not found: %v\n", err)
+	}
+	s, err := generate(flag.Arg(0))
+	if err != nil {
+		log.Fatal(err)
+	}
+	err = gccRun(s)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func gccRun(program string) error {
+	f, err := os.CreateTemp("", "go-crypto-openssl-*.c")
+	if err != nil {
+		log.Fatal(err)
+	}
+	name := f.Name()
+	if !*work {
+		defer os.Remove(name)
```

Comment:
How will `os.Remove` respond when there is an open file handle? On Windows such an operation would typically result in an error.

Reply:
`os.Remove` is a thin wrapper around `DeleteFileW`, so it will also fail.

In this particular case we don't care if there is an error, at the end it is a file in %TEMP% that we try to remove just to be good citizens and don't overfill the temp folder.